src/ray/raylet/scheduling/local_resource_manager.cc

```diff
@@ -393,6 +392,48 @@ void LocalResourceManager::OnResourceOrStateChanged() {
   resource_change_subscriber_(ToNodeResources());
 }
 
+rpc::NodeDeathInfo LocalResourceManager::DeathInfoFromDrainRequest() {
+  rpc::NodeDeathInfo death_info;
+  RAY_CHECK(drain_request_ != nullptr);
+  if (drain_request_->reason() ==
+      rpc::autoscaler::DrainNodeReason::DRAIN_NODE_REASON_IDLE_TERMINATION) {
+    death_info.set_reason(rpc::NodeDeathInfo::AUTOSCALER_DRAIN_IDLE);
+    death_info.set_reason_message(drain_request_->reason_message());
+  } else {
+    RAY_CHECK_EQ(drain_request_->reason(),
+                 rpc::autoscaler::DrainNodeReason::DRAIN_NODE_REASON_PREEMPTION);
+    death_info.set_reason(rpc::NodeDeathInfo::AUTOSCALER_DRAIN_PREEMPTED);
+    death_info.set_reason_message(drain_request_->reason_message());
+  }
+  return death_info;
+}
+
+rpc::NodeDeathInfo LocalResourceManager::AdjustDeathInfo(
+    const rpc::NodeDeathInfo &death_info) {
+  if (death_info.reason() != rpc::NodeDeathInfo::EXPECTED_TERMINATION) {
+    return death_info;
+  }
+  if (drain_request_ == nullptr) {
+    RAY_LOG(DEBUG) << "Skip adjusting node death info because there is no existing "
+                      "drain request.";
+    return death_info;
+  } else if (drain_request_->reason() !=
+                 rpc::autoscaler::DrainNodeReason::DRAIN_NODE_REASON_PREEMPTION ||
+             drain_request_->deadline_timestamp_ms() == 0 ||
+             drain_request_->deadline_timestamp_ms() > current_time_ms()) {
+    RAY_LOG(DEBUG)
+        << "Skip adjusting node death info because the existing drain request "
+           "would not be the cause of current termination. Existing drain request: "
+        << drain_request_->DebugString();
+    return death_info;
+  }
+
+  rpc::NodeDeathInfo adjusted_death_info;
+  adjusted_death_info.set_reason(rpc::NodeDeathInfo::AUTOSCALER_DRAIN_PREEMPTED);
+  adjusted_death_info.set_reason_message(drain_request_->reason_message());
+  return adjusted_death_info;
+}
```

Comment:
I think it's better to have the logic inside the sigterm handler so the adjustment is only limited there. We can let local_resource_manager, node_manager, raylet exposes the ongoing drain request to main.cc, e.g `std::optional<DrainRequest> GetLocalNodeDrainRequest()`

Reply:
Good idea. Updated.